### PR TITLE
feat: try impl masked mha for sm90a

### PR DESCRIPTION
**Pull request by leavelet**
# Add Sparse Mask Support for Masked Multi-Head Attention

## Summary
This PR adds support for fine-grained sparse masks in FlashAttention, enabling efficient topK-based sparse attention patterns on Hopper (sm90a) architecture.

Mask generation kernel: https://github.com/leavelet/sparse_mask_lib

### 1. **Sparse Mask Infrastructure**
- Added `sparse_mask_fine` parameter: 3D tensor `[total_q, max_k_blocks, num_int32_per_block]` containing block-level bitmaps
- Supports fine-grained masking at the token level with efficient block-based storage
- Requires 128-byte TMA alignment for optimal performance

### 2. **New API Components**

**C++/CUDA:**
- Added sparse mask parameters to `Flash_fwd_params` structure
- Implemented TMA-based mask loading pipeline (`MainloopPipelineMask`)
- Added `SparseMask` class in `mask.h` with optimized GMMA-friendly bit layout
- Added build-time flag `FLASHATTENTION_DISABLE_SPARSE_MASK` for optional compilation

**Python:**
- Added `sparse_mask_fine` parameter to all forward functions:
  - `flash_attn_func()`
  - `flash_attn_varlen_func()`
  - `flash_attn_with_kvcache()`
- Added `get_tile_size()` utility function to query `(kBlockM, kBlockN)` for mask generation

### 3. **Implementation Details**

**Mask Layout Optimization:**
- Reordered bitmap layout for optimal GMMA access patterns
- For `kBlockN=128`: 4 words, 32 bits per quad-lane
- For `kBlockN=64`: 2 words, 16 bits per quad-lane
- Achieves zero memory bank conflicts during mask application

**Pipeline Integration:**
- Dual-path support: TMA (primary) and cp.async (fallback)
- Mask loading overlapped with K/V loading in producer warp
- Consumer threads apply masks during score computation with minimal overhead

### 4. **Constraints & Requirements**
- Hopper architecture (SM90+) only
- Row stride must be multiple of 128 bytes for TMA alignment
- When sparse mask is enabled, causal masking is disabled

## Usage Example

```python
import torch
from flash_attn_interface import flash_attn_varlen_func, get_tile_size
from sparse_mask import prepare_sparse_mask

# Query tile size for mask generation
kBlockM, kBlockN = get_tile_size(
            headdim=layer.head_dim,
            headdim_v=layer.v_head_dim,
            qkv_dtype=q.dtype,
            is_causal=False,  # causality handled by topk_indices
        )

# Generate sparse mask (e.g., from topK selection)
num_int32_per_block = (kBlockN + 31) // 32
max_k_blocks = (max_seqlen_k + kBlockN - 1) // kBlockN
sparse_mask = torch.zeros(total_q, max_k_blocks, num_int32_per_block, 
                          dtype=torch.int32, device='cuda')
# Generate sparse masks from topk_indices
total_q = q.shape[0]
max_seqlen_k = metadata.max_seq_len_k
# Pad max_seqlen_k to be multiple of 16 for TMA load

max_k_blocks = (max_seqlen_k + kBlockN - 1) // kBlockN
num_int32_per_block = (kBlockN + 31) // 32
product = max_k_blocks * num_int32_per_block
if product % 32 != 0:
    aligned_product = ((product + 31) // 32) * 32
    max_k_blocks = aligned_product // num_int32_per_block

fine_mask = prepare_sparse_mask(
            topk_indices=topk_indices,
            total_q=total_q,
            max_seqlen_k=max_seqlen_k,
            max_k_blocks=max_k_blocks,
            kBlockN=kBlockN,
            kBlockM=kBlockM,
            reordered=True,
        )

# Run attention with sparse mask
out = flash_attn_varlen_func(
    q, k, v,
    cu_seqlens_q, cu_seqlens_k,
    max_seqlen_q, max_seqlen_k,
    sparse_mask_fine=sparse_mask
)
```

## Testing
- Validated on H100 with various sequence lengths and head dimensions

### CODE REVIEW

**Comment by Fridge003** [comment]
So amazing!

**Comment by FlamingoPg** [comment]
Looks good!

**Comment by Fridge003** [comment]
Hi @leavelet , can you please submit this PR again with the SM80 issue fixed, thanks

**Comment by leavelet** [comment]
> Hi @leavelet , can you please submit this PR again with the SM80 issue fixed, thanks

I will fix that next week
